### PR TITLE
Finish up Astrometric frame and spherical_offsets_to 

### LINKED ISSUE #4209: Function or method to compute the RA and Dec offsets between two coordinates?
This might be a naive question, but I could not find the answer in the documentation. Is there a function or method to compute the RA and Dec offsets between two coordinates? I'm looking for a more direct way of doing this:

```python
import numpy as np
from astropy.coordinates import SkyCoord

mm1 = SkyCoord('03h28m55.58s', '+31d14m37.06s')
mm2 = SkyCoord('03h28m55.71s', '+31d14m35.33s')

ra_offset = (mm2.ra - mm1.ra) * np.cos(mm1.dec.to('radian'))
dec_offset = (mm2.dec - mm1.dec)
print (ra_offset.to('arcsec'), dec_offset.to('arcsec'))
```

EDIT by @eteq: 
The discussion below has lead to two follow on issues, which when closed/merged will mean we can close this:
- [ ] #4338 (`spherical_offset` method)
- [ ] #4339 (`AstrometricFrame`)

### LINKED ISSUE #4339: Create AstrometricFrame
This issue is the second part of #4209, along side #4338 : this issue is about creating a new frame (tentatively called `AstrometricFrame`) that is a rotation of some other frame to be centered at a specific point.

From one my comments in #4209 proposing this:
> Also implement a new frame  - lets call it ``AstrometricFrame``?  This would be a frame with one frame attribute: the "origin" coordinate.  This would along the lines of what @mhvk and @smaret are proposing, although it has to be a full rotation (rather than just ``cos(dec)`` flattening) to account for positions that might be farther away than the small angle approximation allows. Note that the code to do this is *already* in the `GalactocentricFrame` transformation - that's actually even *more* complicated than this because it allows an additional "roll" angle (whereas here we're saying the new y-axis has to point in the same direction as the old y-axis at the astrometric point).  We could also add a `SkyCoord` convenience method to get the astrometric frame of a coordinate, allowing use like ``coord1.transform_to(coord2.astrometric_frame())``, which would then trivially give us the angles we'd need for `spherical_offsets`.

Perhaps @mhvk, @smaret, or @taldcroft would be interested in implementing this (based on the #4209 conversation)?  Although certainly other volunteers are welcome...

### PR DESCRIPTION
**Pull request by eteq**
This PR is an amalgamation of related work, particularly from #4338 and #4524 (it replaces both).  Thanks to @alexrudy and @Kevtron for their contributions in the latter!

I won't dwell on the ideas too much because they been covered in all the other issues I reference here and have been reviewed piece-meal in the various PRs.  So if the tests pass my feeling is that this is all set!

One significant change from #4524 that's worth mentioning though: this now uses some metaclass magic to convince the class generator (which creates astrometric classes) to both give correct class names (previously they were all called `Astrometric` which causes various issues - now they get names like `AstrometricICRS`), and change the spherical components to e.g. `dra` and `ddec` / `dlon` and `dlat` (instead of `ra` and `dec` or `lon` and `lat`).  I think this is safer because that makes it immediately clear when reading the code that it's an astrometric ("offset") frame instead of just an ordinary frame.

So if merged, this closes #4209, closes #4338, closes #4339, and closes #4524 (whew!)

### CODE REVIEW

**Comment by eteq** [comment]
(Oh, and it assuages @r-own and @mhvk's concerns from #4338 by no longer using the small-angle approximation: `spherical_offsets_to` is now basically just a shell to simplify the use of astrometric frames for the one-to-one use case)

File: astropy/coordinates/tests/test_astrometric_transformations.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
+# -*- coding: utf-8 -*-
```
Comment (mhvk) [inline astropy/coordinates/tests/test_astrometric_transformations.py]: Duplicate lines...

File: astropy/coordinates/tests/test_astrometric_transformations.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        actual = icrs_coord.transform_to(astrometric_frame)
+        actual_xyz = actual.cartesian.xyz
+
+        # back to ICRS
+        roundtrip = actual.transform_to(ICRS)
+        roundtrip_xyz = roundtrip.cartesian.xyz
+
+        # Verify
+        assert_allclose(actual_xyz.to(u.kpc), expected_xyz.to(u.kpc), atol=1E-5*u.kpc)
+        #assert_allclose(actual_xyz.to(u.kpc), roundtrip_xyz.to(u.kpc), atol=1E-5*u.kpc)
```
Comment (mhvk) [inline astropy/coordinates/tests/test_astrometric_transformations.py]: Why is the commented out? Should it be `astrometric_frame.data.xyz.to(u.kpc), roundtrip...`

Comment (mhvk) [inline astropy/coordinates/tests/test_astrometric_transformations.py]: Or just removed? It isn't there for `dec` either.

File: astropy/coordinates/tests/test_sky_coord.py [hunk 3]
```diff
@@ -1171,3 +1170,39 @@ def test_getitem_representation():
...
+    assert 0*u.arcmin < dra < 1*u.arcmin
+
+    fk5 = SkyCoord(0*u.arcmin, 0*u.arcmin, frame='fk5')
+
+    with pytest.raises(ValueError):
+        # different frames should fail
+        i00.spherical_offsets_to(fk5)
+
+
+    #large offset (>1 deg) should give a warning
```
Comment (mhvk) [inline astropy/coordinates/tests/test_sky_coord.py]: No longer true, so remove this comment!

**Comment by mhvk** [comment]
@eteq - this looks great, although the actual machinery is still beyond me (it would be good to have a "coordinates for developpers" section in the documentation, just like @astrofrog has been thinking about a similar one for modelling).

**Comment by mhvk** [comment]
The two remaining failures in travis seem to real (it is a doctest).

File: astropy/coordinates/tests/test_astrometric_transformations.py (hunk 2 above)
Comment (eteq) [inline astropy/coordinates/tests/test_astrometric_transformations.py]: Oops, yeah, I think that's supposed to be removed (this was from the #4524 branch and I didn't look through it terribly carefully...)

**Comment by eteq** [comment]
Alright, I think I've addressed all of @mhvk's comments.  

Re: 
>  the actual machinery is still beyond me (it would be good to have a "coordinates for developpers" section in the documentation, just like @astrofrog has been thinking about a similar one for modelling).

Good idea!  I created #4913 to try to help us remember to do this.  I admit there's some real black magic going on in the metaclasses here.  Although I confess I enjoyed it: there's some really tricky stuff going on there that is really only possible due to Python's inherent flexibility.

File: astropy/coordinates/tests/test_astrometric_transformations.py (hunk 1 above)
Comment (eteq) [inline astropy/coordinates/tests/test_astrometric_transformations.py]: (I did fix this, it's just not obvious because git didn't hide this line)

**Comment by eteq** [comment]
weird... the remaining failure is due to doctests that come out as either ``0`` or ``-0`` at machine precision.  I would have thought FLOAT_CMP would fix that but apparently not... at any rate, I can just change the examples to do something that's not quite so close to 0.

**Comment by mhvk** [comment]
Sounds good to just change the example. I think this reflects a display issue in older versions of numpy -- I'd guess FLOAT_CMP would pick it up if there were a dot after the number (which there should be).

**Comment by eteq** [comment]
in the end I concluded that it was good to keep one example with 0s (to demonstrate the idea) and just skip that in the doctests.  The rest are all non-0 now (and otherwise essentially identical), so it's still sufficiently tested.

**Comment by taldcroft** [comment]
@eteq - rather late to this conversation, but I had always assumed that astrometric frames would include a roll (or parallactic) angle.  This has practical application as a poor-mans WCS to make a quick sky image (perhaps a finding chart) based on a catalog and telescope parallactic angle.

**Comment by eteq** [comment]
@taldcroft -Ah, good point.  Was pretty easy to include, so the commit I just pushed up includes a `rotation` attribute (which defaults to 0).

One catch with this: it's now somewhat less right to call it "dra" or "ddec" if the rotation is non-0.  But I think it's ok to just accept that - it's still reasonably clear, and one can always do ``coord.spherical.lat`` or similar.

**Comment by taldcroft** [comment]
@eteq - awesome on getting in the rotation!  I had expected to have to wait until 1.3.  :smile:

**Comment by eteq** [comment]
Oh, one thing about the rotation: I just picked the sign in the way that I thought was the most likely use case ("rotate to match a thing at position angle x"), but I'm really not sure if that's the most intuitive direction.  Probably best to just merge this once the tests are passing, and then have an issue about that, as we can gather opinions during beta.

**Comment by eteq** [comment]
Turns out there's a really obscure bug in py 3.3: the ``namedtuple._asdict`` method does not work correctly if a namedtuple is a subclass of another namedtuple.  My only question is: who was it that found this and fixed it in 3.4?? This is a pretty deep obscure hole in the python world...

Anyway, after the last commit (which just adds a compatibility layer to fix up the py 3.3 asdict), we should be good to go.

**Comment by astrofrog** [comment]
Looks good!

**Comment by mhvk** [comment]
@eteq - on the names: from what I've seen from, e.g., POSS and UCAC calibration, if one goes for relative positions, sometimes xi and eta are used as local longitude and latitude relative to a plate centre (e.g., the UCAC paper by Zacharias et al. http://adsabs.harvard.edu/abs/2000AJ....120.2131Z). But those of course still have unknown position, so arguably that would be a bit different astrometric frame. Still, it may not be bad to use those for the names, since for large values, `dra` and `ddec` are not offsets in ra, dec any more (dra=12h, ddec=0 would get one to ra_ref+12h, -dec_ref). And it doesn't have problems at the poles...

As for the rotation, I think one always it to be such that with rot=0, a star to the North from the ref star would have a positive offset in latitude, with rot=90, that would hold for a star in the East, etc., i.e., the rotation defines where a positive offset in latitude goes, from N going through E.

**Comment by eteq** [comment]
(I've created #4931 to continue the conversation since this is now merged as-is)

**Comment by mhvk** [comment]
@eteq - the doctest problem with `-0.` and `0.` was because `FLOAT_CMP` did not use the sign of numbers; see https://github.com/astropy/astropy/pull/4920/commits/41293bfa4eac067dd53f0530fd2b71829894f3e3

**Pull request merged**